File: NMSMB Scripts/CMKushnir/Portal.cs
//=============================================================================
// Make all runes known, all portal buttons free (no charge substance required).
//=============================================================================

namespace cmk.NMS.Scripts.Mod
{
	using CompletionRequirementEnum = GcMaintenanceElement.CompletionRequirementEnum;
	
	//=========================================================================
	
	public class Portal: cmk.NMS.ModScript
	{
		protected override void Execute()
		{
			Try(() => GcGameplayGlobals());
			Try(() => TkAttachmentData());
		}

		//...........................................................

		protected void GcGameplayGlobals()
		{
			var mbin = ExtractMbin<GcGameplayGlobals>(
				"GCGAMEPLAYGLOBALS.GLOBAL.MBIN"
			);
			// know all portal glyphs, see Start.
			mbin.Portal.KnowAllRunes = true;
		}
		
		//...........................................................

		protected void TkAttachmentData()
		{
			var mbin = ExtractMbin<TkAttachmentData>(
				"MODELS/PLANETS/BIOMES/COMMON/BUILDINGS/PORTAL/PORTAL/ENTITIES/BUTTON.ENTITY.MBIN"
			);
			// make all portal buttons free (don't need to charge w/ substances)
			var maint = mbin.Components.FindFirst<GcMaintenanceComponentData>();
			//foreach( var button in maint.PreInstalledTech ) {
			//	button.MaxCapactiy           = 0;    // -1, old way
			//	button.MinRandAmount         = 100;  //  0, old old way
			//	button.MaxRandAmount         = 100;  //  0, old old way
			//	button.CompletionRequirement = CompletionRequirementEnum.NoRequirement;  // new way
			//}
			// another option is to delete all maint comp data for the buttons
			mbin.Components.Remove(maint);
		}
	}
}

//=============================================================================

File: other-authors/Portal.lua
NMS_MOD_DEFINITION_CONTAINER = {
["MOD_BATCHNAME"]   = "cmk.pak",
["MOD_FILENAME"]    = "Portal.pak",
["MOD_DESCRIPTION"] = "Disable charge requirements for portals",
["MOD_AUTHOR"]      = "cmk",
["NMS_VERSION"]     = "3.21.0",
["MODIFICATIONS"]   = {
{
  ["MBIN_CHANGE_TABLE"] = { 
  {
    ["MBIN_FILE_SOURCE"]  = "MODELS/PLANETS/BIOMES/COMMON/BUILDINGS/PORTAL/PORTAL/ENTITIES/BUTTON.ENTITY.MBIN",
    ["EXML_CHANGE_TABLE"] = {
    {
      ["REPLACE_TYPE"]       = "ALL",                  -- make all 16 buttons no-cost
      ["VALUE_CHANGE_TABLE"] = {{"MaxCapacity", "0"}}  -- -1, 0 makes button  no-cost
    }}
  }}  -- MBIN_CHANGE_TABLE
}}  -- MODIFICATIONS
}  -- NMS_MOD_DEFINITION_CONTAINER
